x-pack/filebeat/input/gcppubsub/input.go

```diff
@@ -209,30 +213,70 @@ func (in *pubsubInput) run() error {
 	ctx, cancel := context.WithCancel(in.workerCtx)
 	defer cancel()
 
-	client, err := in.newPubsubClient(ctx)
-	if err != nil {
-		return err
+	numGoRoutines := in.Subscription.NumGoroutines
+	var workerWg sync.WaitGroup
+
+	for ctx.Err() == nil {
+
+		workers, err := in.workerSem.AcquireContext(numGoRoutines, ctx)
+		if err != nil {
+			break
+		}
+
+		if workers == 0 {
+			continue
+		}
+		workerWg.Add(workers)
+		for i := 0; i < workers; i++ {
+			go func() {
+				client, err := in.newPubsubClient(ctx)
+				defer func() {
+					workerWg.Done()
+					in.workerSem.Release(1)
```

Comment:
Switch order; we're not done until we've communicated state.

Reply:
Switched the order